### ISSUE TITLE
Auto discover and manual not working

### ISSUE THREAD

**Issue by kmahelona** [calvarium/homebridge-lifx-plugin#9]
**Describe The Bug:**
Auto discover AND manual IP of bulbs doesn’t work for me. I could be missing something. My two Lifx A19 (night vision) lights are on the network, they’re connected well (I have a Unifi network) with fixed IP addresses.

**To Reproduce:**

1. Ensure your Lifx A19 are connected to the network. They are, but they are not accessible via Lifx App OR HomeKit... and that's the main reason why I'm using this app.
2. Install the plugin.
3. Wait for auto discover (nothing happens)
4. Add the IP addresses of the two lights (nothing happens). I don't see anything in the accessories and the logs I get are 
```
DEBUG - 24000034285761bb00000000000000000000000000000000000000000000000002000000 to 192.168.1.190
DEBUG - 24000034285761bb00000000000000000000000000000000000000000000000002000000 to 192.168.1.11
DEBUG - 24000034285761bb00000000000000000000000000000000000000000000000002000000 to 255.255.255.255
```

**Expected behavior:**
I'd expect the lights to show up somewhere but tbh I'm not sure where/how -- improved documentation could help. 

Some questions I have are?
1. What is the broadcast address? If my router's subnet mask is 255.255.255.0, then should I change the broadcast address to that (I've tried that and nothing)?
2. What is the bind address? Is it important? 0.0.0.0 to me sounds like "accept connections from all IPs" so I'm guessing that's fine.
3. Are devices meant to show up automatically in the Accessories? That appears to be the case for the Unifi plugin that I have.
4. Am I supposed to add devices found with this plugin manually to HomeKit or do I just add the bridge and then they show up automatically in HomeKit (sorry I'm a noob here so this might be a silly question.)

**Logs:**

```
[02/02/2022, 20:40:19] Homebridge v1.4.0 (HAP v0.10.0) (Homebridge FE6E) is running on port 51336.
[02/02/2022, 20:40:20] [homebridge-unifi-protect] Launched child bridge with PID 9832
[02/02/2022, 20:40:20] [Lifx Plugin] Launched child bridge with PID 9833
[02/02/2022, 20:40:20] Registering platform 'homebridge-lifx-plugin.LifxPlugin'
[02/02/2022, 20:40:20] [Lifx Plugin] Loaded homebridge-lifx-plugin v0.1.20 child bridge successfully
[02/02/2022, 20:40:20] Loaded 0 cached accessories from cachedAccessories......
[02/02/2022, 20:40:20] Homebridge v1.4.0 (HAP v0.10.0) (Lifx Plugin) is running on port 41724.
DEBUG - 24000034285761bb00000000000000000000000000000000000000000000000002000000 to 255.255.255.255
DEBUG - 24000034285761bb00000000000000000000000000000000000000000000000002000000 to 192.168.1.190
DEBUG - 24000034285761bb00000000000000000000000000000000000000000000000002000000 to 192.168.1.11
[02/02/2022, 20:40:21] Registering platform 'homebridge-unifi-protect.UniFi Protect'
[02/02/2022, 20:40:21] [homebridge-unifi-protect] Loaded homebridge-unifi-protect v5.0.6 child bridge successfully
[02/02/2022, 20:40:21] Loaded 1 cached accessories from cachedAccessories......
[02/02/2022, 20:40:21] Homebridge v1.4.0 (HAP v0.10.0) (homebridge-unifi-protect) is running on port 53458.
[02/02/2022, 20:40:21] [homebridge-unifi-protect] UCK-G2-Plus [UCK-G2-PLUS]: Connected to the UniFi Protect controller API (address: 192.168.1.36 mac: ....).
[02/02/2022, 20:40:21] [homebridge-unifi-protect] UCK-G2-Plus [UCK-G2-PLUS]: Connected to the UniFi Protect realtime update events API.
[02/02/2022, 20:40:21] [homebridge-unifi-protect] UCK-G2-Plus [UCK-G2-PLUS]: Discovered camera: Entrance [UVC G4 Bullet] (address: 192.168.1.41 mac: ...).
[02/02/2022, 20:40:21] [homebridge-unifi-protect] UCK-G2-Plus [UCK-G2-PLUS]: Controller refresh interval set to 10 seconds.
[02/02/2022, 20:40:21] [homebridge-unifi-protect] UCK-G2-Plus [UCK-G2-PLUS] Entrance [UVC G4 Bullet]: Smart motion detection enabled: person, vehicle.
[02/02/2022, 20:40:21] [homebridge-unifi-protect] UCK-G2-Plus [UCK-G2-PLUS] Entrance [UVC G4 Bullet]: Smart motion contact sensors enabled: person, vehicle.
DEBUG - 24000034285761bb00000000000000000000000000000000000000000000000002000000 to 255.255.255.255
DEBUG - 24000034285761bb00000000000000000000000000000000000000000000000002000000 to 192.168.1.190
DEBUG - 24000034285761bb00000000000000000000000000000000000000000000000002000000 to 192.168.1.11
DEBUG - 24000034285761bb00000000000000000000000000000000000000000000000002000000 to 255.255.255.255
DEBUG - 24000034285761bb00000000000000000000000000000000000000000000000002000000 to 192.168.1.190
DEBUG - 24000034285761bb00000000000000000000000000000000000000000000000002000000 to 192.168.1.11
DEBUG - 24000034285761bb00000000000000000000000000000000000000000000000002000000 to 255.255.255.255
DEBUG - 24000034285761bb00000000000000000000000000000000000000000000000002000000 to 192.168.1.190
DEBUG - 24000034285761bb00000000000000000000000000000000000000000000000002000000 to 192.168.1.11
DEBUG - 24000034285761bb00000000000000000000000000000000000000000000000002000000 to 255.255.255.255
DEBUG - 24000034285761bb00000000000000000000000000000000000000000000000002000000 to 192.168.1.190
DEBUG - 24000034285761bb00000000000000000000000000000000000000000000000002000000 to 192.168.1.11
DEBUG - 24000034285761bb00000000000000000000000000000000000000000000000002000000 to 255.255.255.255
DEBUG - 24000034285761bb00000000000000000000000000000000000000000000000002000000 to 192.168.1.190
DEBUG - 24000034285761bb00000000000000000000000000000000000000000000000002000000 to 192.168.1.11
DEBUG - 24000034285761bb00000000000000000000000000000000000000000000000002000000 to 255.255.255.255
DEBUG - 24000034285761bb00000000000000000000000000000000000000000000000002000000 to 192.168.1.190
DEBUG - 24000034285761bb00000000000000000000000000000000000000000000000002000000 to 192.168.1.11
DEBUG - 24000034285761bb00000000000000000000000000000000000000000000000002000000 to 255.255.255.255
DEBUG - 24000034285761bb00000000000000000000000000000000000000000000000002000000 to 192.168.1.190
DEBUG - 24000034285761bb00000000000000000000000000000000000000000000000002000000 to 192.168.1.11
DEBUG - 24000034285761bb00000000000000000000000000000000000000000000000002000000 to 255.255.255.255
DEBUG - 24000034285761bb00000000000000000000000000000000000000000000000002000000 to 192.168.1.190
DEBUG - 24000034285761bb00000000000000000000000000000000000000000000000002000000 to 192.168.1.11
DEBUG - 24000034285761bb00000000000000000000000000000000000000000000000002000000 to 255.255.255.255
DEBUG - 24000034285761bb00000000000000000000000000000000000000000000000002000000 to 192.168.1.190
DEBUG - 24000034285761bb00000000000000000000000000000000000000000000000002000000 to 192.168.1.11
DEBUG - 24000034285761bb00000000000000000000000000000000000000000000000002000000 to 255.255.255.255
DEBUG - 24000034285761bb00000000000000000000000000000000000000000000000002000000 to 192.168.1.190
DEBUG - 24000034285761bb00000000000000000000000000000000000000000000000002000000 to 192.168.1.11
DEBUG - 24000034285761bb00000000000000000000000000000000000000000000000002000000 to 255.255.255.255
DEBUG - 24000034285761bb00000000000000000000000000000000000000000000000002000000 to 192.168.1.190
DEBUG - 24000034285761bb00000000000000000000000000000000000000000000000002000000 to 192.168.1.11
DEBUG - 24000034285761bb00000000000000000000000000000000000000000000000002000000 to 255.255.255.255
DEBUG - 24000034285761bb00000000000000000000000000000000000000000000000002000000 to 192.168.1.190
DEBUG - 24000034285761bb00000000000000000000000000000000000000000000000002000000 to 192.168.1.11
DEBUG - 24000034285761bb00000000000000000000000000000000000000000000000002000000 to 255.255.255.255
DEBUG - 24000034285761bb00000000000000000000000000000000000000000000000002000000 to 192.168.1.190
DEBUG - 24000034285761bb00000000000000000000000000000000000000000000000002000000 to 192.168.1.11
DEBUG - 24000034285761bb00000000000000000000000000000000000000000000000002000000 to 255.255.255.255
DEBUG - 24000034285761bb00000000000000000000000000000000000000000000000002000000 to 192.168.1.190
DEBUG - 24000034285761bb00000000000000000000000000000000000000000000000002000000 to 192.168.1.11
DEBUG - 24000034285761bb00000000000000000000000000000000000000000000000002000000 to 255.255.255.255
DEBUG - 24000034285761bb00000000000000000000000000000000000000000000000002000000 to 192.168.1.190
DEBUG - 24000034285761bb00000000000000000000000000000000000000000000000002000000 to 192.168.1.11
DEBUG - 24000034285761bb00000000000000000000000000000000000000000000000002000000 to 255.255.255.255
DEBUG - 24000034285761bb00000000000000000000000000000000000000000000000002000000 to 192.168.1.190
DEBUG - 24000034285761bb00000000000000000000000000000000000000000000000002000000 to 192.168.1.11
DEBUG - 24000034285761bb00000000000000000000000000000000000000000000000002000000 to 255.255.255.255
DEBUG - 24000034285761bb00000000000000000000000000000000000000000000000002000000 to 192.168.1.190
DEBUG - 24000034285761bb00000000000000000000000000000000000000000000000002000000 to 192.168.1.11
DEBUG - 24000034285761bb00000000000000000000000000000000000000000000000002000000 to 255.255.255.255
DEBUG - 24000034285761bb00000000000000000000000000000000000000000000000002000000 to 192.168.1.190
DEBUG - 24000034285761bb00000000000000000000000000000000000000000000000002000000 to 192.168.1.11

```

**Plugin Config:**

```json
{
    "bridge": {
        "name": "Homebridge FE6E",
        "username": "—-",
        "port": 51336,
        "pin": "—-"
    },
    "accessories": [],
    "platforms": [
        {
            "name": "Config",
            "port": 8581,
            "platform": "config"
        },
        {
            "controllers": [
                {
                    "address": "192.168.1.36",
                    "password": "—",
                    "username": "homebridge"
                }
            ],
            "platform": "UniFi Protect",
            "_bridge": {
                "username": "—-",
                "port": 53458
            },
            "options": [
                "Enable.Motion.SmartDetect",
                "Enable.Motion.SmartDetect.ObjectSensors"
            ]
        },
        {
            "name": "Lifx Plugin",
            "duration": 0,
            "brightnessDuration": 300,
            "colorDuration": 300,
            "default": "0.0.0.0",
            "broadcast": "255.255.255.255",
            "lightOfflineTolerance": 3,
            "messageHandlerTimeout": 45000,
            "resendPacketDelay": 150,
            "resendMaxTimes": 3,
            "debug": true,
            "autoDiscover": true,
            "bulbs": [
                {
                    "name": "light 1",
                    "address": "192.168.1.190"
                },
                {
                    "name": "light 2",
                    "address": "192.168.1.11"
                }
            ],
            "_bridge": {
                "username": "—-",
                "port": 41724
            },
            "platform": "LifxPlugin"
        }
    ],
    "disabledPlugins": []
}
```

**Screenshots:**

**Environment:**
I’ve got a RaspberryPi running Homebridge. I’ve got a mixed 5GHz 2.4GHz wifi (I sometimes have to disable the 5Ghz to get the Lifx bulbs to connect). FWIW I’ve struggled to get one of these Lifx bulbs to connect (though I did mange eventually) hence why I’m trying trying this plugin. Then idk HomeKit couldn’t find the lifx lights despite their being connected to the network (always verifiable through UniFi’s nice mobile app) … I did unplug my HomePod and move it around a bit. So that’s the other part fo this environment is the HomePod is the “base station”.

I'm gonna try and reset these all again and see if I can get them to at least load on Lifx app. But got it's such a pain. The features of these lights are great but I think 50% of the time I struggle to connect them. And it doesn't help that the HomeKit code is so small on the light. Do I need to keep QR codes around for ever?

* **Plugin Version**: 0.1.20
* **Homebridge Version**: 1.4.0
* **Node.js Version**: 16.13.2
* **NPM Version**: 8.3.1
* **Operating System**: Raspbian GNU/Linux Bullseye (11)

**Comment by calvarium** [comment]
Hello @kmahelona,

after reading your report I think that the plugin can't find your devices, because it searches in a different network range than you have. Also I recommend you to adjust the parameters broadcast address and default route to your ip-range.

Maybe the explanation to your questions can bring light into the darkness.

**1. What is the broadcast address?**

A broadcast address is a network address used for transmission to all devices connected to a multiple-access communication network. A message sent to a broadcast address can be received by all hosts connected to the network. 
This will probably also be your problem, because the plugin in the current configuration does not use the correct broadcast address and therefore cannot communicate with all devices.

**2. What is the bind address? Is it important?** 

I am not 100% sure but I think you mean the default route. You can read about it in READ.me. This is an entry in a routing table to which all packets are forwarded for which there is no explicit entry in the table. Means if your Pi would route in different networks you could determine herewith which network it should address.
In your case I would use the default configuration as I don't expect your Pi to send to 2 networks.


**3. Are devices meant to show up automatically in the Accessories? That appears to be the case for the Unifi plugin that I have.**

Yes, all devices that can communicate over the network with the broadcast address should immediately be listed as accessories. If this is not the case, it indicates that the broadcast address or default route is not correct if you have autodiscover enabled.
Also, manually adding the individual IP's is not necessary as they should all be detected automatically.

**4. Am I supposed to add devices found with this plugin manually to HomeKit or do I just add the bridge and then they show up automatically in HomeKit**

Once a wise man said: "There are no silly questions, only stupid answers".
You just need to add your bridge to Homekit, your bulbs will all be added automatically. Adding them manually is not necessary.


In summary, my answer is:
First find out your correct broadcast address and adjust your default route if necessary. But I think that the correct broadcast address will be enough.

As a little hint I give you the command
`ifconfig`
Just execute it in the terminal on your Pi.
In the second line of the output you will find your current broadcast address at the end:
`inet xxx.xxx.xxx.xxx netmask xxx.xxx.xxx.xxx broadcast 255.255.255.255`

I'll think about how I can improve my documentation, but I hope I could help you.
Let me know if everything worked for you

**Comment by kmahelona** [comment]
Thanks. That worked. It was the broadcast address and your tip was exactly what I needed. I'm happy to open a PR for the README to help others who might run into the same problem I had. Alternatively maybe a startup script could run the `ifconfig` and write that as the configuration?

I still do have a light that can't be found but I think that's Lifx's fault and not the plugin 😅

**Comment by gvimalku** [comment]
I have a similar issue, the Homebridge UI says the plug in is successfully installed, I updated the broadcast address based on my ifconfig response, but doesn't show up in the Accessories section of the Homebridge UI, let alone Homekit. I'm using a Raspberry Pi 4 for Homebridge. I've left the default/bind ip as it is.
